cirq/sim/clifford/clifford_gate_decomposer.py

```diff
@@ -0,0 +1,73 @@
+from typing import (Tuple)
+
+import numpy as np
+
+from cirq import protocols, ops
+
+# All 24 Clifford gates, written as products of H and S gates.
+_ALL_CLIFFORD_GATES = [
+    '', 'HSSH', 'HSSSH', 'HSH', 'SS', 'HSSHSS', 'SHSSS', 'SSSHS', 'HSSS',
+    'SHSSSH', 'S', 'HSSHSSS', 'HS', 'SSHSSS', 'SSS', 'HSSHS', 'H', 'SSH', 'SH',
+    'HSHS', 'HSS', 'SSHSS', 'SHSS', 'SSSHSS'
+]
+
+
+class CliffordGateDecomposer:
+
+    def __init__(self):
+        s = protocols.unitary(ops.S)
+        h = protocols.unitary(ops.H)
+
+        # Converts string made of letters S and H to gate.
+        def name_to_unitary(name):
+            result = np.eye(2)
+            for ch in name:
+                if ch == 'S': result = result @ s
+                elif ch == 'H': result = result @ h
+            return result
+
+        # Calculate matrices for all candidates.
+        self._candidates = [
+            (name, name_to_unitary(name)) for name in _ALL_CLIFFORD_GATES
```

Comment:
every time a CliffordgateDecomposer is initialized, this map is reconstructed. It would be more efficient to have it as a hardcoded constant map instead.

Reply:
Yes, that's why I made it a singleton, so this computation can be done once in runtime.
This computation is not that expensive - just 2x2 matrix multiplications ~100 times.

I refactored this code to use static methods instead, so now it's clear that this computation is done only once.